doc/designs/quic-design/connection-id-cache.md

```diff
@@ -0,0 +1,257 @@
+
+/*
+ * Query a remote connection for a connection ID.
+ * Each connection can have multiple connection IDs associated with different
+ * routes.  This function returns one of these in a non-specified manner.
+ */
+int ossl_quic_conn_id_cache_get0_remote_conn_id(
+        const QUIC_REMOTE_CONN_ID_CACHE *cache,
+        const QUIC_CONNECTION *conn, QUIC_CONN_ID **cid);
+
+/*
+ * Retire remote connection IDs up to and including the one determined by the
+ * sequence number.
+ */
+int ossl_quic_conn_id_cache_retire_remote_conn_ids(
+        QUIC_REMOTE_CONN_ID_CACHE *cache, uint64_t seq_no);
+
+/*
+ * Remove remote connection IDs up to and including the one determined by the
+ * sequence number.
+ */
+int ossl_quic_conn_id_cache_retire_remove_conn_ids(
+        QUIC_REMOTE_CONN_ID_CACHE *cache, uint64_t seq_no);
+```
+
+### Local Connection ID APIs
+
+```c
+typedef struct quic_local_conn_id_cache_st QUIC_LOCAL_CONN_ID_CACHE;
+
+/*
+ * Allocate and free a local connection ID cache
+ */
+QUIC_LOCAL_CONN_ID_CACHE *ossl_quic_local_conn_id_cache_new_conn(void);
+void ossl_quic_local_conn_id_cache_free(QUIC_LOCAL_CONN_ID_CACHE *cache);
+
+/*
+ * Generate a new random local connection ID and associate it with a connection.
+ * For MVP this could just be a zero length ID.
+ */
+int ossl_quic_local_conn_id_cache_new_conn_id(QUIC_LOCAL_CONN_ID_CACHE *cache,
+                                              QUIC_CONNECTION *conn,
+                                              QUIC_CONN_ID **cid);
+
+/*
+ * Remove a local connection and all associated cached IDs
+ */
+int ossl_quic_remove_connection(QUIC_LOCAL_CONN_ID_CACHE *cache,
+                                const QUIC_CONNECTION *conn);
+
+/*
+ * Lookup a local connection by ID.
+ * Returns the connection or NULL if absent.
+ */
+QUIC_CONNECTION *ossl_quic_get0_local_conn(
+        const QUIC_LOCAL_CONN_ID_CACHE *cache,
+        const unsigned char *conn_id, size_t conn_id_len);
+
+/*
+ * Retire local connection IDs up to and including the one specified by the
+ * sequence number.
+ */
+int ossl_quic_conn_id_cache_retire_local_conn_ids(
+        QUIC_LOCAL_CONN_ID_CACHE *cache, uint64_t from_seq_no);
+
+/*
+ * Remove local connection IDs up to and including the one specified by the
+ * sequence number.
+ */
+int ossl_quic_conn_id_cache_remove_local_conn_ids(
+        QUIC_LOCAL_CONN_ID_CACHE *cache, uint64_t from_seq_no);
+```
+
+### Routes
+
+Additional status and source information is also included.
+
+```c
+typedef struct quic_route_st QUIC_ROUTE;
+typedef struct quic_route_table QUIC_ROUTE_TABLE;
+
+typedef struct quic_route_st {
+    QUIC_CONN_ID local;
+    QUIC_CONN_ID remote;
+    uint64_t     seq_no;        /* Sequence number for both ends */
+    unsigned int retired : 1;   /* Connection ID has been retired */
+#if 0
+    /* Later will require */
+    BIO_ADDR remote_address;    /* remote source address */
+    /* which exposes union bio_addr_st more widely */
+#endif
+};
+```
```

Comment:
Needs API for instantiating routing table, any other operations, etc. How are routes and routing tables created and what operations can be performed on them?